### DISCUSSION TITLE
Identifying SQL within strings of another language

### DISCUSSION THREAD

**Discussion by tinygiant98** [tree-sitter/tree-sitter#2835] [Q&A]
A while back, I created a tree sitter grammar for a custom language we use. The grammar works well. We've recently added the ability to interface with sqlite from this language. I'd like to inject sql grammar, but it should only be injected into nodes that are identified as sql queries.

Here's the challenge I've been unable to overcome...

The queries are written as any other query, but are contained within double quoted strings. They can also be concatenated, just like any string.

Here's a simple example: 
```
string q = "BEGIN TRANSACTION;";
```
and a slightly more complicated example: 
```
string q = "SELECT * FROM " + sTable + " WHERE ...;";
```

We would then send that string into the sqlite engine for processing. The implementation works well, but for syntax coloring and parsing, is there a way to identify this string as an sql string versus a normal string, and then label it as an sql node so the sql grammar can be injected to parse it? 

Notably, we'd want to be able to identify the string `string s = "BEGIN"` as a normal non-sql string versus when it's used within an sql query.

One method we're considering is adding a prefix/qualifier to the string, like `string q = q"BEGIN TRANSACTION;";`  to explicitly identify queries, but that may still not resolve the issue with concatenated queries.

Thanks!

**Comment by Fapp3r3Lit3** [comment]
Is this for a certain Tap to P script? Through either iPhone or droid? Inbox

**Reply by tinygiant98** [reply]
It is not for mobile. Primarily for syntax highlighting in vscode, I'm trying to identify sql statements which are contained within otherwise normal double quoted strings.  The catch, of course, is that these strings can be concatenated, but I'll cross that bridge when I get to it.  I'm primarily interested in a method to identify sql so I can mark that node for grammar injection.

**Comment by ObserverOfTime** [comment]
Pro tip: never concatenate SQL queries with arbitrary variables.

**Reply by tinygiant98** [reply]
Understood, however, at the time of my original question, my (closed) environment didn't allow the use of variable binding for sqlite, so concatenation was the only option.  Variable binding is allowed now, so no inline concatenation is used anymore.
